src/curvefit/models/core_model.py

```diff
@@ -151,11 +110,113 @@ def objective(self, x, data):
             param_gprior=self.data_inputs.param_gprior_info,
         ) 
 
+    def gradient(self, x, data):
+        if self.data_inputs is None:
+            self.data_inputs = convert_inputs(self.param_set, data)
+        finfo = np.finfo(float)
+        step  = finfo.tiny / finfo.eps
+        x_c = x + 0j
+        grad = np.zeros(x.size)
+        for i in range(x.size):
+            x_c[i] += step*1j
+            grad[i] = self.objective(x_c, data).imag/step
+            x_c[i] -= step*1j
+
+        return grad
+
+    def forward(self, x, t):
+        params = effects2params(
+            x, 
+            self.data_inputs.group_sizes, 
+            self.data_inputs.covariates_matrices,
+            self.param_set.link_fun,
+            self.data_inputs.var_link_fun,
+        )
+        return self.curve_fun(t, params[:, 0])
+
```

Comment:
Can add a `predict` method that takes `x`, `t`, `group`, and function that can be different than `self.curve_fun`.

```
def predict(self, x, t, group, predict_fun=None):
    params = effects2params(...)
    if predict_fun is None:
        predict_fun = self.curve_fun
    return predict_fun(t, params[:, 0]) # not sure how group gets in here with the current structure -- thoughts?

Reply:
yes, I'll added that! right now the model gets group info from data and store it in `data_inputs`. would we be fitting data on one group and predicting on another? if not I think maybe we can just not pass in group.

Reply:
True... so maybe we actually don't need that! The only case where we would need it is if we make group-specific predictions for a model that had multiple groups in the fitting. But I don't actually think we do that anymore anywhere.

Reply:
We do need the ability to have a different `predict_fun` though

Reply:
yes, we should definitely allow different `predict_fun`.